drivers/display/display_framebuf.c

```diff
@@ -0,0 +1,251 @@
+	}
+}
+
+static int framebuf_set_orientation(const struct device *dev,
+				    const enum display_orientation orientation)
+{
+	switch (orientation) {
+	case DISPLAY_ORIENTATION_NORMAL:
+		return 0;
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static void framebuf_get_capabilities(const struct device *dev,
+				      struct display_capabilities *caps)
+{
+	struct framebuf_dev_data *data;
+
+	data = FRAMEBUF_DATA(dev);
+	caps->x_resolution = data->width;
+	caps->y_resolution = data->height;
+	caps->supported_pixel_formats = PIXEL_FORMAT_ARGB_8888 |
+					PIXEL_FORMAT_MONO01 |
+					PIXEL_FORMAT_MONO10;
+	caps->screen_info = 0;
+	caps->current_pixel_format = data->format;
+	caps->current_orientation = DISPLAY_ORIENTATION_NORMAL;
+}
+
+/*
+ * simple bit manipulation functions for display buffer read/write.
+ *
+ * we can't use the system-defined ones because they work on words,
+ * not bytes, which would introduce a dependency on endianness.
+ */
+
+static inline int test_bit(const void *buf, int bit)
+{
+	const u8_t *buf8 = buf;
+
+	return buf8[bit >> 3] & (1 << (7 - (bit & 7)));
+}
+
+static inline void set_bit(void *buf, int bit)
+{
+	u8_t *buf8 = buf;
+
+	buf8[bit >> 3] |= (1 << (7 - (bit & 7)));
+}
+
+static inline void reset_bit(void *buf, int bit)
+{
+	u8_t *buf8 = buf;
+
+	buf8[bit >> 3] &= ~(1 << (7 - (bit & 7)));
+}
+
+static int framebuf_write(const struct device *dev, const u16_t x,
+			  const u16_t y,
+			  const struct display_buffer_descriptor *desc,
+			  const void *buf)
+{
+	struct framebuf_dev_data *data = FRAMEBUF_DATA(dev);
+	u32_t *dst = data->buffer;
+	const u32_t *src;
+	int row, col, bit;
+	u32_t set_color, reset_color;
+
+	dst += x;
+	dst += (y * data->pitch);
+
+	switch (data->format) {
+	case PIXEL_FORMAT_ARGB_8888:
+		src = buf;
+		for (row = 0; row < desc->height; ++row) {
+			memcpy(dst, src, desc->width * sizeof(u32_t));
+			dst += data->pitch;
+			src += desc->pitch;
+		}
+		break;
+	case PIXEL_FORMAT_MONO01:
+	case PIXEL_FORMAT_MONO10:
+		if (data->format == PIXEL_FORMAT_MONO01) {
+			set_color = GREY;
+			reset_color = BLACK;
+		} else {
+			set_color = BLACK;
+			reset_color = GREY;
+		}
+
+		bit = 0;
+		for (row = 0; row < desc->height; ++row) {
+			for (col = 0; col < desc->width; ++col, ++bit) {
+				if (test_bit(buf, bit)) {
+					dst[col] = set_color;
+				} else {
+					dst[col] = reset_color;
+				}
+			}
+			dst += data->pitch;
+			bit += (desc->pitch - col);
+		}
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int framebuf_read(const struct device *dev, const u16_t x,
+			 const u16_t y,
+			 const struct display_buffer_descriptor *desc,
+			 void *buf)
+{
+	struct framebuf_dev_data *data = FRAMEBUF_DATA(dev);
+	u32_t *src = data->buffer;
+	u32_t *dst;
+	int row, col, bit;
+
+	src += x;
+	src += (y * data->pitch);
+
+	switch (data->format) {
+	case PIXEL_FORMAT_ARGB_8888:
+		dst = buf;
+		for (row = 0; row < desc->height; ++row) {
+			memcpy(dst, src, desc->width * sizeof(u32_t));
```

Comment:
misra-C pedantry: need to cast these memcpy calls here and in other places to void since we ignore return value

Reply:
fixed